cmd/lotus-chainwatch/processor/processor.go

```diff
@@ -0,0 +1,309 @@
+package processor
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/filecoin-project/specs-actors/actors/builtin"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/lib/parmap"
+)
+
+var log = logging.Logger("processor")
+
+type Processor struct {
+	db *sql.DB
+
+	node api.FullNode
+
+	// number of blocks processed at a time
+	batch int
+}
+
+type ActorTips map[types.TipSetKey][]actorInfo
+
+type actorInfo struct {
+	act types.Actor
+
+	stateroot cid.Cid
+	height    abi.ChainEpoch // so that we can walk the actor changes in chronological order.
+
+	tsKey       types.TipSetKey
+	parentTsKey types.TipSetKey
+
+	addr  address.Address
+	state string
+}
+
+func NewProcessor(db *sql.DB, node api.FullNode, batch int) *Processor {
+	return &Processor{
+		db:    db,
+		node:  node,
+		batch: batch,
+	}
+}
+
+func (p *Processor) setupSchemas() error {
+	if err := p.setupMiners(); err != nil {
+		return err
+	}
+
+	if err := p.setupRewards(); err != nil {
+		return err
+	}
+
+	if err := p.setupMessages(); err != nil {
+		return err
+	}
+
+	if err := p.setupCommonActors(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Processor) Start(ctx context.Context) {
+	log.Info("Starting Processor")
+
+	if err := p.setupSchemas(); err != nil {
+		log.Fatalw("Failed to setup processor", "error", err)
+	}
+
+	go p.subMpool(ctx)
+
+	// main processor loop
+	go func() {
+		for {
```

Comment:
This `for` should handle ctx.Done and exit.

Reply:
In `main.go` we (somewhat ungracefully) call `os.Exit()` when `ctx.Done()` returns. Would there be any value added by implementing a graceful termination here in light of this?

Reply:
If we somehow change the app in a way that the context ends but the app continues running, this goroutine will never resolve. Someone who's moving the consumption of processer.Start shouldn't have to know there's an immortal goroutine living within that now needs to be handled properly. They see `Start(ctx context.Context)` and assume the context is respected (and rightfully so).